contrib/mpi-proxy-split/mpi-wrappers/p2p-deterministic.c

```diff
@@ -0,0 +1,152 @@
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <mpi.h>
+// To support MANA_P2P_LOG and MANA_P2P_REPLAY:
+#define USE_READALL
+#define USE_WRITEALL
+#include "p2p-deterministic.h"
+
+/***********************************************************
+ * Utitilies for point-to-point deterministic log-and-replay
+ ***********************************************************/
+
+static struct p2p_log_msg next_msg_entry;
+static struct p2p_log_msg *next_msg = NULL;
+
+void p2p_log(int count, MPI_Datatype datatype, int source, int tag,
+             MPI_Comm comm, MPI_Status *status, MPI_Request *request) {
+  if (status) {
+    source = status->MPI_SOURCE;
+    tag = status->MPI_TAG;
+    int count;
+    MPI_Get_count(status, MPI_CHAR, &count);
+    datatype = MPI_CHAR;
+  }
+  set_next_msg(count, datatype, source, tag, comm, NULL, request);
+}
+
+void initialize_next_msg(int fd) {
+  next_msg = &next_msg_entry;
+  readall(fd, next_msg, sizeof(*next_msg));
+}
+
+int iprobe_next_msg(struct p2p_log_msg *p2p_msg) {
+  /* FIXME:  This isn't comiling yet.
+  if (!next_msg) {
+    initialize_next_msg(fd);
+  }
+  */
+  if (p2p_msg) {
+    *p2p_msg = next_msg_entry;
+  }
+  return (next_msg_entry.comm != MPI_COMM_NULL);
+}
+
+void get_next_msg(struct p2p_log_msg *p2p_msg) {
+  static int fd = -2;
+  if (fd == -2) {
+    char buf[100];
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    snprintf(buf, sizeof(buf)-1, P2P_LOG_MSG, rank);
+    fd = open(buf, O_RDONLY);
+    if (fd == -1) {
+      perror("get_next_msg: open");
+      exit(1);
+    }
+  }
+  if (!next_msg) {
+    initialize_next_msg(fd);
+  }
+  *p2p_msg = next_msg_entry;
+  readall(fd, next_msg, sizeof(*next_msg));
+}
+
+
+// comm defined
+// Either status and request are non-null, or source, tag defined.
+void set_next_msg(int count, MPI_Datatype datatype,
+                  int source, int tag, MPI_Comm comm,
+                  MPI_Status *status, MPI_Request *request) {
+  struct p2p_log_msg p2p_msg;
+  static int fd = -2;
+  if (fd == -2) {
+    char buf[100];
+    int rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    snprintf(buf, sizeof(buf)-1, P2P_LOG_MSG, rank);
+    fd = open(buf, sizeof(buf)-1,
+              O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR );
```

Comment:
Please remove the second argument "sizeof(buf)-1" from open call.

Reply:
Done.  Great catch again.  I don't know why the compiled didn't pick this up.